File: app/src/main/java/com/sduduzog/slimlauncher/data/AppRepository.java
package com.sduduzog.slimlauncher.data;

import android.app.Application;
import android.arch.lifecycle.LiveData;
import android.os.AsyncTask;

import java.util.List;

public class AppRepository {
    private AppDao appDao;
    private LiveData<List<App>> apps;

    public AppRepository(Application application){
        AppRoomDatabase db = AppRoomDatabase.getDatabase(application);
        appDao = db.appDao();
        apps = appDao.getAllApps();
    }

    public LiveData<List<App>> getAllApps() {
        return apps;
    }

    public void insert(App app){
        new insertAsyncTask(appDao).execute(app);
    }

    public void delete(String packageName) {
        new deleteAsyncTask(appDao).execute(packageName);
    }

    private static class insertAsyncTask extends AsyncTask<App, Void, Void> {

        private AppDao mAsyncTaskDao;

        insertAsyncTask(AppDao dao) {
            mAsyncTaskDao = dao;
        }

        @Override
        protected Void doInBackground(final App... params) {
            mAsyncTaskDao.insert(params[0]);
            return null;
        }
    }

    private static class deleteAsyncTask extends AsyncTask<String, Void, Void> {

        private AppDao mAsyncTaskDao;

        deleteAsyncTask(AppDao dao) {
            mAsyncTaskDao = dao;
        }

        @Override
        protected Void doInBackground(final String... params) {
            mAsyncTaskDao.delete(params[0]);
            return null;
        }
    }
}

File: app/src/main/java/com/sduduzog/slimlauncher/data/AppRepository.kt
package com.sduduzog.slimlauncher.data

import android.app.Application
import android.os.AsyncTask
import androidx.lifecycle.LiveData
import com.sduduzog.slimlauncher.data.model.HomeApp

class AppRepository(application: Application) {
    private val db: DataRoomDatabase = DataRoomDatabase.getDatabase(application)!!
    private val appDao: AppDao = db.appDao()

    private val _apps = appDao.homeApps

    val apps: LiveData<List<HomeApp>>
        get() = _apps

    fun add(app: HomeApp){
        AddAppAsyncTask(appDao).execute(app)
    }

    fun update(vararg list: HomeApp) {
        UpdateAppAsyncTask(appDao).execute(*list)
    }

    fun remove(app: HomeApp){
        RemoveAppAsyncTask(appDao).execute(app)
    }

    private class AddAppAsyncTask(private val mAsyncTaskDao: AppDao) : AsyncTask<HomeApp, Void, Void>() {

        override fun doInBackground(vararg params: HomeApp): Void? {
            mAsyncTaskDao.add(params[0])
            return null
        }
    }

    private class UpdateAppAsyncTask(private val mAsyncTaskDao: AppDao) : AsyncTask<HomeApp, Void, Void>() {

        override fun doInBackground(vararg params: HomeApp): Void? {
            mAsyncTaskDao.update(*params)
            return null
        }
    }

     private class RemoveAppAsyncTask(private val mAsyncTaskDao: AppDao) : AsyncTask<HomeApp, Void, Void>() {

        override fun doInBackground(vararg params: HomeApp): Void? {
            mAsyncTaskDao.remove(*params)
            return null
        }
    }
}